pkg/cmd/drtprod/cli/commands/yamlprocessor.go

```diff
@@ -137,56 +140,114 @@ func setEnv(environment map[string]string, displayOnly bool) error {
 func processTargets(
 	ctx context.Context, targets []target, displayOnly bool, targetNames []string,
 ) error {
+	// targetNameMap is used to check all targets that are provided as user input
 	targetNameMap := make(map[string]struct{})
-	targetMap := make(map[string][]*command)
 	for _, tn := range targetNames {
 		targetNameMap[tn] = struct{}{}
 	}
-	for i := 0; i < len(targets); i++ {
-		targets[i].TargetName = os.ExpandEnv(targets[i].TargetName)
-		t := targets[i]
-		if _, ok := targetNameMap[t.TargetName]; len(targetNames) > 0 && !ok {
-			fmt.Printf("Ignoring execution for target %s\n", t.TargetName)
-			continue
-		}
-		// Generate the commands for each target's steps
-		targetSteps, err := generateCmdsFromSteps(t.TargetName, t.Steps)
-		if err != nil {
-			return err
-		}
-		targetMap[t.TargetName] = targetSteps
+	waitGroupTracker, err := buildTargetCmdsAndRegisterWaitGroups(targets, targetNameMap, targetNames)
+	if err != nil {
+		return err
 	}
 
-	// Use a WaitGroup to execute commands concurrently
+	// if displayOnly, we just print and exit
+	if displayOnly {
+		for _, t := range targets {
+			if shouldProcessTarget(targetNameMap, t, targetNames) {
+				continue
+			}
+			displayCommands(t)
+		}
+		return nil
+	}
+	// Use a WaitGroup to wait for commands executed concurrently
 	wg := sync.WaitGroup{}
-	for targetName, cmds := range targetMap {
-		if displayOnly {
-			displayCommands(targetName, cmds)
+	for _, t := range targets {
+		if shouldProcessTarget(targetNameMap, t, targetNames) {
 			continue
 		}
 		wg.Add(1)
-		go func(tn string, commands []*command) {
-			err := executeCommands(ctx, tn, commands)
+		go func(t target) {
+			// defer complete the wait group for the dependent targets to proceed
+			defer waitGroupTracker[t.TargetName].Done()
+			defer wg.Done()
+			for _, dt := range t.DependentTargets {
+				if twg, ok := waitGroupTracker[dt]; ok {
+					fmt.Printf("%s: waiting on <%s>\n", t.TargetName, dt)
+					// wait on the dependent targets
+					// it would not matter if we wait sequentially as all dependent targets need to complete
+					twg.Wait()
+				}
+			}
+			err := executeCommands(ctx, t.TargetName, t.commands)
 			if err != nil {
-				fmt.Printf("%s: Error executing commands: %v\n", tn, err)
+				fmt.Printf("%s: Error executing commands: %v\n", t.TargetName, err)
 			}
-			wg.Done()
-		}(targetName, cmds)
+		}(t)
 	}
+	// final wait for all targets to complete
 	wg.Wait()
 	return nil
 }
 
+// shouldProcessTarget returns true if the target should be selected for processing
+func shouldProcessTarget(targetNameMap map[string]struct{}, t target, targetNames []string) bool {
```

Comment:
I think you meant to flip the condition? You can either do that or rename the function to `shouldSkipTarget` and update the comment

```
    if shouldSkipTarget(targetNameMap, t, targetNames) {
			continue
		}
```

Reply:
We're continuing continuing everywhere the condition is true. Renaming seems simpler.

Reply:
Thanks or catching this! Done!